src/analysis/processing/qgsalgorithmlinedensity.cpp

```diff
@@ -0,0 +1,193 @@
+  mSearchGeometry = firstCellMidpoint.buffer( mSearchRadius, 5 );
+  mSearchGeometryArea = mDa.measureArea( mSearchGeometry );
+
+  mIndex = QgsSpatialIndex( *mSource, nullptr, QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+
+  return true;
+}
+
+QVariantMap QgsLineDensityAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  int rows = std::max( std::ceil( mExtent.height() / mPixelSize ), 1.0 );
+  int cols = std::max( std::ceil( mExtent.width() / mPixelSize ), 1.0 );
+
+  //build new raster extent based on number of columns and cellsize
+  //this prevents output cellsize being calculated too small
+  QgsRectangle rasterExtent = QgsRectangle( mExtent.xMinimum(), mExtent.yMaximum() - ( rows * mPixelSize ), mExtent.xMinimum() + ( cols * mPixelSize ), mExtent.yMaximum() );
+
+  std::unique_ptr< QgsRasterFileWriter > writer = qgis::make_unique< QgsRasterFileWriter >( outputFile );
+  writer->setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer->createOneBandRaster( Qgis::Float32, cols, rows, rasterExtent, mCrs ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+
+  provider->setNoDataValue( 1, -9999 );
+
+  int totalCellcnt = rows * cols;
+  int cellcnt = 0;
+
+  for ( int row = 0; row < rows; row++ )
+  {
+    std::unique_ptr< QgsRasterBlock > rasterDataLine = qgis::make_unique< QgsRasterBlock >( Qgis::Float32, cols, 1 );
+    for ( int col = 0; col < cols; col++ )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+
+      if ( col > 0 )
+        mSearchGeometry.translate( mPixelSize, 0 );
+
+      QList<QgsFeatureId> fids = mIndex.intersects( mSearchGeometry.boundingBox() );
+
+      std::unique_ptr< QgsGeometryEngine > engine( QgsGeometry::createGeometryEngine( mSearchGeometry.constGet() ) );
+      engine->prepareGeometry();
+
+      //remove non-intersecting fids
+      int i = 0;
+      while ( i < fids.size() )
+      {
+        QgsFeatureId fid = fids.at( i );
+        QgsGeometry lineGeom = mIndex.geometry( fid );
+
+        if ( engine->intersects( lineGeom.constGet() ) == false )
+        {
+          fids.removeAt( i );
+        }
+        i++;
+      }
+
+      QgsFeatureIds isectingFids = QSet< QgsFeatureId >::fromList( fids );
+      QgsFeatureIterator fit = mSource->getFeatures( QgsFeatureRequest().setFilterFids( isectingFids ) );
```

Comment:
This is potentially quite costly (firing off a feature request for every pixel). I'd suggest you could avoid this, by building up a hash of the weights vs feature id at the same time you populate the spatial index. That means only a single full iteration of the layer is required, instead of potentially 1000's.